### PR TITLE
oto: remove Cgo on darwin

### PR DESCRIPTION
**Pull request by TotallyGamerJet**
I accidentally deleted my fork. This is a new PR for removing Cgo on darwin. It works with TotallyGamerJet/purego objc branch

### CODE REVIEW

**Comment by TotallyGamerJet** [comment]
Do you think that the structs that are for `AudioToolbox.framework` should be in `purego/frameworks/AudioToolbox` package or `purego/objc/frameworks/AudioToolbox`? Obviously, we wouldn't provide all functions nor structures for the the frameworks just the one's Oto or Ebitengine need. That way we only Dlopen the frameworks once using `import "framework..."` instead of every time someone wants to use that framework.

**Comment by hajimehoshi** [comment]
> Do you think that the structs and or functions that are for AudioToolbox.framework should be in purego/frameworks/AudioToolbox package or purego/objc/frameworks/AudioToolbox? Obviously, we wouldn't provide all functions nor structures for the the frameworks just the one's Oto or Ebitengine need. That way we only Dlopen the frameworks once using import "framework..." instead of every time someone wants to use that framework.

I don't think we should provide purego/frameworks/AudioToolbox so far. It's because we will have more responsibility to maintain the AudioToolbox API. The less public APIs, the happier we are. Let's revisit this when we really want this.

**Comment by TotallyGamerJet** [comment]
@hajimehoshi this is good for a review

File: driver_ios.go [hunk 1]
```diff
@@ -22,3 +22,11 @@ package oto
 // '4' is float32 size in bytes. '2' is a number of channels for stereo.
 
 const bufferSizeInBytes = 12288
+
+var _ = purego.Dlopen("/System/Library/Frameworks/AppKit.framework/Versions/Current/AppKit", purego.RTLD_GLOBAL)
```
Comment (hajimehoshi) [inline driver_ios.go:26]: Why do we need this line?

File: api_audio_darwin.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
+package oto
+
+import _ "runtime/cgo" // TODO: remove once purego(#12) is solved.
+import (
+	"unsafe"
+
+	"github.com/ebitengine/purego"
+)
```
Comment (hajimehoshi) [inline api_audio_darwin.go:8]: I prefer one import section:

```go
import (
	_ "runtime/cgo" // TODO: remove once purego(#12) is solved.
	"unsafe"

	"github.com/ebitengine/purego"
)
```

File: api_audio_darwin.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	toolbox                    = purego.Dlopen("/System/Library/Frameworks/AudioToolbox.framework/Versions/Current/AudioToolbox", purego.RTLD_GLOBAL)
+	atAudioQueueNewOutput      = purego.Dlsym(toolbox, "AudioQueueNewOutput")
+	atAudioQueueStart          = purego.Dlsym(toolbox, "AudioQueueStart")
+	atAudioQueuePause          = purego.Dlsym(toolbox, "AudioQueuePause")
+	atAudioQueueAllocateBuffer = purego.Dlsym(toolbox, "AudioQueueAllocateBuffer")
+	atAudioQueueEnqueueBuffer  = purego.Dlsym(toolbox, "AudioQueueEnqueueBuffer")
+)
+
+func _AudioQueueNewOutput(inFormat *_AudioStreamBasicDescription, inCallbackProc _AudioQueueOutputCallback, inUserData unsafe.Pointer, inCallbackRunLoop uintptr, inCallbackRunLoopMod uintptr, inFlags uint32, outAQ *_AudioQueueRef) uintptr {
+	ret, _, _ := purego.SyscallN(atAudioQueueNewOutput,
```
Comment (hajimehoshi) [inline api_audio_darwin.go:70]: So we can ignore errors from SyscallN, right?

File: driver_macos.go [hunk 4]
```diff
@@ -17,7 +17,50 @@
...
+	"github.com/ebitengine/purego"
+	"github.com/ebitengine/purego/objc"
+)
 
 const bufferSizeInBytes = 2048
+
+var appkit = purego.Dlopen("/System/Library/Frameworks/AppKit.framework/Versions/Current/AppKit", purego.RTLD_GLOBAL)
+
+// oto_setNotificationHandler sets a handler for sleep/wake notifications.
+func oto_setNotificationHandler() {
```
Comment (hajimehoshi) [inline driver_macos.go:32]: We no longer need the prefix `oto_` for a function

File: driver_macos.go [hunk 5]
```diff
@@ -17,7 +17,50 @@
...
+	// Create the Observer object
+	class := objc.AllocateClassPair(objc.GetClass("NSObject\x00"), "OtoNotificationObserver\x00", 0)
+	class.AddMethod(objc.RegisterName("receiveSleepNote:\x00"), objc.IMP(oto_setGlobalPause), "v@:@\x00")
+	class.AddMethod(objc.RegisterName("receiveWakeNote:\x00"), objc.IMP(oto_setGlobalResume), "v@:@\x00")
+	class.Register()
+	//OtoNotificationObserver *observer = [OtoNotificationObserver new];
```
Comment (hajimehoshi) [inline driver_macos.go:38]: Is it intended to leave comments for the original ObjC code?

File: driver_darwin.go [hunk 6]
```diff
@@ -14,60 +14,50 @@
...
-	if osstatus := C.AudioQueueNewOutput(
+	var audioQueue _AudioQueueRef
+	if osstatus := _AudioQueueNewOutput(
 		&desc,
-		(C.AudioQueueOutputCallback)(C.oto_render),
+		oto_render,
 		nil,
-		(C.CFRunLoopRef)(0),
-		(C.CFStringRef)(0),
+		0, //CFRunLoopRef
```
Comment (hajimehoshi) [inline driver_darwin.go:49]: Add a space after `//`

File: driver_ios.go (hunk 1 above)
Comment (TotallyGamerJet) [reply driver_ios.go:26]: I don't think we do

File: api_audio_darwin.go (hunk 3 above)
Comment (TotallyGamerJet) [reply api_audio_darwin.go:70]: We can since frameworks don't return errors in the standard way

File: driver_macos.go (hunk 5 above)
Comment (TotallyGamerJet) [reply driver_macos.go:38]: It was but I don't think it's needed if you'd rather have them removed

Comment (hajimehoshi) [reply driver_macos.go]: Thanks for removing them!

**Comment by hajimehoshi** [comment]
I'm testing your PR on iOS. Please wait for a while.

**Comment by hajimehoshi** [comment]
I encountered this error with Ebitengine's example (audio) on macOS:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4149487]

goroutine 17 [running, locked to thread]:
github.com/ebitengine/purego.callbackWrap(0x0)
        /Users/hajimehoshi/go/pkg/mod/github.com/ebitengine/purego@v0.0.0-20220724152741-1491ef70dbf7/syscall_darwin.go:108 +0x67
exit status 2
```

Could you take a look? Thanks,

**Comment by TotallyGamerJet** [comment]
I will take a look! Thanks for testing

**Comment by TotallyGamerJet** [comment]
Is it happening for you on amd64 and arm64 or just one of them?

**Comment by TotallyGamerJet** [comment]
I created a pull request on [purego#13](https://github.com/ebitengine/purego/pull/13) that fixes it on amd64. I continue to find no issue with arm64. Are you?

**Comment by hajimehoshi** [comment]
Yes, I tested only on amd64, not arm64.

**Comment by TotallyGamerJet** [comment]
Ok, well then it should all be working then

**Comment by hajimehoshi** [comment]
Thanks. I have confirmed Ebitengine's audio example worked well. Now I'm testing an iOS application.

**Comment by hajimehoshi** [comment]
I found this error on an iOS application (arm64, of course)

```
purego: fn is nil
ui.(*userInterfaceImpl).run.func1()              
        @v2.4.0-alpha.9ui/ui_mobile.go:271 +0x40 
panic({0x100f58ae0, 0x1010077f0})                
        runtime/panic.go:838 +0x204              
github.com/ebitengine/purego.SyscallN(0x100f77500?, {0x130532e00?, 0x1377f3158?, 0x100a9f180?})
        github.com/ebitengine/puregocdc2949de6/syscall.go:25 +0xb4
github.com/hajimehoshi/oto/v2._AudioQueueNewOutput(0x31d8?, 0x10e876b50?, 0x1377f31e8?, 0x0, 0x0, 0x0, 0x1377f31d8?)
        github.com/hajimehoshi/oto/v2@v2.2.0-alpha.6/api_audio_darwin.go:84 +0xa8
github.com/hajimehoshi/oto/v2.newAudioQueue(0xbb80, 0x2, 0x20df1?)
        github.com/hajimehoshi/oto/v2@v2.2.0-alpha.6/driver_darwin.go:45 +0xc8
github.com/hajimehoshi/oto/v2.newContext(0xbb80, 0x2, 0x2)
        github.com/hajimehoshi/oto/v2@v2.2.0-alpha.6/driver_darwin.go:101 +0x144
github.com/hajimehoshi/oto/v2.NewContext(0x1377b2f00?, 0x1008745f0?, 0x1377f3358?)
        github.com/hajimehoshi/oto/v2@v2.2.0-alpha.6/driver.go:100 +0x20
/audio.newContext(0x12?, 0x12?, 0x1377f3378?)
        @v2.4.0-alpha.9/audio/pl
```

**Comment by TotallyGamerJet** [comment]
Well that’s strange. I’ll look into it

**Comment by TotallyGamerJet** [comment]
I'm not sure what is wrong since I am not able to test because go mobile errors with: `gomobile: -target="ios" requires Xcode`. I updated dlopen to be more generalized and not use a specific version. if that doesn't work then I can add logging support to purego that will generate an error when dlopen or dlsym returns 0.

**Comment by hajimehoshi** [comment]
I can test if you add some printfs or logs.

**Comment by hajimehoshi** [comment]
I have confirmed the latest commit fa002c32d34eb80002887c8a35a42a5e92c3b0ec worked on my iPhone SE. I'll test the previous commit again just in case.

**Comment by hajimehoshi** [comment]
The previous commit 71f072f20513f9782fe60d9ffca0619b87e2cfe2 failed with the same error, so this is deterministic.

**Review by hajimehoshi — APPROVED** [review]
LGTM, thank you!

I'll merge this if there is nothing more you want to experiment.

**Comment by TotallyGamerJet** [comment]
I don’t think it is necessary. The issue appears to be that the wrong file path for the framework was used. Perhaps we still go ahead and add dlerror to purego to make it easier to debug issues like these in the future

**Pull request merged**